upload-api/stores/provisions.js

```diff
@@ -0,0 +1,155 @@
+import {
+  QueryCommand,
+  PutItemCommand,
+  DescribeTableCommand,
+} from '@aws-sdk/client-dynamodb'
+import { Failure } from '@ucanto/server'
+import { marshall, } from '@aws-sdk/util-dynamodb'
+
+import { ConflictError as ConsumerConflictError } from '../tables/consumer.js'
+import { ConflictError as SubscriptionConflictError } from '../tables/subscription.js'
+
+class ConflictError extends Failure {
+  /**
+   * @param {object} input
+   * @param {string} input.message
+   */
+  constructor({ message }) {
+    super(message)
+    this.name = 'ConflictError'
+  }
+}
+
+/**
+ * @param {import('../types').SubscriptionTable} subscriptionTable
+ * @param {import('../types').ConsumerTable} consumerTable
+ * @param {import('@ucanto/interface').DID<'web'>[]} services
+ * @returns {import('@web3-storage/upload-api').ProvisionsStorage}
+ */
+export function useProvisionStore (subscriptionTable, consumerTable, services) {
+  return {
+    services,
+    hasStorageProvider: async (consumer) => (
+      { ok: await consumerTable.hasStorageProvider(consumer) }
+    ),
+
+    put: async (item) => {
+      const { cause, consumer, customer, provider } = item
+      // by setting subscription to customer we make it so each customer can have at most one subscription
+      // TODO is this what we want?
```

Comment:
ok - I've updated this to use the CID of `customer`-wrapped-in-a-CBOR-map as the "partition key" and the provider as the  "sort key" (see https://github.com/web3-storage/w3infra/pull/194/files#diff-7f8284bee6fcc2ea7426b91d587ca5eff07fb34f5049b2e95823a03c2768b722) which will give us the uniqueness semantics you describe

re: inserting into both tables I agree, and I think the right path to get there is to break `provider/add` into the two invocations you've talked about (I think `provider/get` and `consumer/add` ?) - I'd like to do that in a separate initiative though since this has gotten quite big and I don't think the failure cases here are too bad to handle for now.